File: src/BuildingBlocks/BuildingBlocks/Result.cs
using System.Text.Json.Serialization;

namespace BuildingBlocks;

[method: JsonConstructor]
public class Result<T>(bool isSuccess, string? error, T? data)
{
    public bool IsSuccess { get; } = isSuccess;

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Error { get; } = error;

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public T? Data { get; } = data;

    private Result(T data) : this(true, null, data)
    {
    }

    private Result(string error) : this(false, error, default(T?))
    {
    }

    public static Result<T> Success(T value) => new(value);
    public static Result<T> Failure(string error) => new(error);
}

File: uis/WebApp/src/app/shared/models/result.ts
export class Result<T> {
  public readonly isSuccess: boolean;
  public readonly error: string | null;
  public readonly data: T | null;

  private constructor(isSuccess: boolean, error: string | null, data: T | null) {
    this.isSuccess = isSuccess;
    this.error = error;
    this.data = data;
  }

  public static success<T>(value: T): Result<T> {
    return new Result(true, null, value);
  }

  public static failure<T>(error: string): Result<T> {
    return new Result<T>(false, error, null as T | null);
  }
}
